File: dashboard/devicetree.js
/* The device tree: several kinds of knowing, rendered so they never look alike.
 *
 * WHAT THIS PANEL IS FOR. The Ethernet scanner above it answers "what is
 * listening?" - a TCP connect and an OUI lookup. That is an INFERENCE. CIP
 * ListIdentity answers "what are you?" - the device's own vendor, product,
 * revision and serial. That is a STATEMENT. Both are useful and they are not
 * the same thing, so every row here says which one it is standing on.
 *
 * The rule this panel follows, and the reason it exists at all: a guess and a
 * statement must never be rendered with the same confidence. It is the same
 * rule as the age on the tag table. A screen that looks equally sure about
 * both is worse than one that shows less, because the operator cannot tell
 * which half to trust and eventually stops trusting either.
 *
 * DISAGREEMENT IS SHOWN, NEVER RESOLVED. If the OUI table and the device give
 * different vendors, both appear and the row is marked. Picking a winner is how
 * the wrong one ends up on screen with nothing to say it was ever in doubt.
 *
 * NOTHING HERE PUTS TRAFFIC ON A WIRE BY ITSELF. The tree is a GET over what is
 * already known. Discovery is a button, because a UDP broadcast onto a plant
 * segment should never happen because somebody opened a tab.
 */
(() => {
  'use strict';
  const root = document.getElementById('devicePanel');
  if (!root) return;

  let api, status, groups, discoverBtn, message;
  let snapshot = null;

  function el(tag, cls, text) {
    const node = document.createElement(tag);
    if (cls) node.className = cls;
    if (text !== undefined && text !== null) node.textContent = text;
    return node;
  }

  function build() {
    root.replaceChildren();
    const intro = el('p', null,
      'The Ethernet scan and CIP discovery, merged. Every row says which of them '
      + 'reported it: a device that answered ListIdentity told you what it is, one '
      + 'with only an open port did not.');
    root.appendChild(intro);

    status = el('p');
    status.className = 'field-state';
    status.setAttribute('aria-live', 'polite');
    root.appendChild(status);

    const bar = el('div', 'row');
    discoverBtn = el('button', 'btn', 'Discover (CIP broadcast)');
    discoverBtn.title = 'Sends one UDP ListIdentity broadcast on this segment and '
      + 'records what answers. Read-only: ListIdentity has no write form.';
    discoverBtn.addEventListener('click', () => onDiscover());
    bar.appendChild(discoverBtn);
    root.appendChild(bar);

    message = el('p', 'muted', '');
    message.setAttribute('aria-live', 'polite');
    root.appendChild(message);

    groups = el('div', 'device-groups');
    root.appendChild(groups);
  }

  // ── the two empty states, which are different answers ──────────────────────
  //
  // "Nothing has scanned yet" and "the scan found nothing" mean opposite things
  // and used to look identical everywhere in this dashboard. An empty list with
  // no explanation reads as the second when it is usually the first.
  function emptyText(snap) {
    const scanned = snap.scan_state && snap.scan_state !== 'idle';
    const discovered = snap.discovery && snap.discovery.at;
    if (!scanned && !discovered) {
      return 'Nothing has looked yet. Run the Ethernet scanner above, or press '
           + 'Discover for a CIP broadcast.';
    }
    return 'Nothing answered. The scan and discovery ran and found no devices on '
         + 'this segment.';
  }

  function sourceChip(source) {
    const chip = el('span', 'src-chip', source);
    chip.dataset.source = source;
    // The tooltip is where the distinction is actually spelled out, because the
    // chip has room for a name and not for a caveat.
    chip.title = {
      'cip-listidentity': 'The device answered a broadcast with its own vendor, '
        + 'product, revision and serial. It told you.',
      'segment-scan': 'A TCP connect succeeded and the MAC resolved through an OUI '
        + 'table. Something is listening and the NIC was sold by that vendor - '
        + 'that is an inference, not the device speaking.',
      'opcua-endpoints': 'The device listed its OPC UA endpoints. Self-reported, '
        + 'and optional: the embedded server is firmware- and SKU-dependent.',
      'promoted': 'You already saved this one to the device list.',
    }[source] || source;
    return chip;
  }

  function renderDevice(device) {
    const row = el('div', 'device-row');
    row.dataset.selfReported = String(device.sources.includes('cip-listidentity')
      || device.sources.includes('opcua-endpoints'));

    const head = el('div', 'device-head');
    head.appendChild(el('span', 'device-addr', device.address));

    const identity = device.identity || {};
    // What it calls itself, or the honest absence of that. Never a blank.
    const name = identity.product_name || device.hostname;
    head.appendChild(el('span', 'device-name', name || 'unidentified'));
    if (!name) {
      head.lastChild.title = 'Nothing reported a product name or a hostname for '
        + 'this address. It is an open port, not a known device.';
    }

    const chips = el('span', 'device-sources');
    for (const source of device.sources) chips.appendChild(sourceChip(source));
    head.appendChild(chips);
    row.appendChild(head);

    const detail = el('div', 'device-detail');
    if (device.vendor) {
      // The vendor ALWAYS travels with where it came from. A vendor with no
      // provenance beside it is the exact thing this panel exists to avoid.
      detail.appendChild(el('span', 'kv', `vendor: ${device.vendor}`));
      detail.appendChild(el('span', 'kv-src', `(${device.vendor_source})`));
    }
    for (const [label, value] of [['rev', identity.revision], ['serial', identity.serial],
                                  ['type', identity.device_type]]) {
      if (value) detail.appendChild(el('span', 'kv', `${label}: ${value}`));
    }
    if (device.ports.length) {
      detail.appendChild(el('span', 'kv',
        'ports: ' + device.ports.map(p => `${p.port}/${p.service}`).join(', ')));
    }
    row.appendChild(detail);

    // Disagreement, shown rather than resolved.
    for (const conflict of device.conflicts) {
      const warn = el('div', 'device-conflict');
      warn.appendChild(el('span', 'conflict-label', `${conflict.field}: sources disagree`));
      for (const entry of conflict.values) {
        warn.appendChild(el('span', 'conflict-value', `${entry.value} — ${entry.source}`));
      }
      row.appendChild(warn);
    }

    if (device.promoted) {
      row.appendChild(el('div', 'device-saved',
        `saved as "${device.promoted.name}" (${device.promoted.protocol || 'no protocol'})`));
    }
    return row;
  }

  function render() {
    if (!snapshot || !status) return;
    const snap = snapshot;
    const scanned = snap.scan_state && snap.scan_state !== 'idle';
    const discovery = snap.discovery || {};

    // The banner says what has actually happened, so an empty tree is never
    // mistaken for an answer about the network.
    const parts = [];
    parts.push(scanned ? `scan: ${snap.scan_state}` : 'scan: not run');
    if (discovery.available === false) {
      parts.push(`CIP discovery unavailable: ${discovery.error || 'unknown'}`);
    } else if (discovery.at) {
      parts.push(`CIP discovery: ${discovery.count} answered`);
    } else {
      parts.push('CIP discovery: not run');
    }
    if (snap.conflicts) parts.push(`${snap.conflicts} disagreement(s)`);
    status.textContent = parts.join(' · ');
    status.dataset.state = snap.conflicts ? 'warn'
      : (snap.total ? 'ok' : 'idle');

    groups.replaceChildren();
    if (!snap.total) {
      groups.appendChild(el('p', 'empty', emptyText(snap)));
      return;
    }
    for (const group of snap.groups) {
      const block = el('div', 'device-group');
      const heading = el('h4', null, group.network);
      // The count that matters: how many told us what they are, versus how many
      // merely answered a port.
      heading.appendChild(el('span', 'group-count',
        `${group.count} found · ${group.self_reported} identified themselves`));
      block.appendChild(heading);
      for (const device of group.devices) block.appendChild(renderDevice(device));
      groups.appendChild(block);
    }
  }

  async function onDiscover() {
    discoverBtn.disabled = true;
    message.textContent = 'Broadcasting ListIdentity…';
    try {
      const response = await fetch('/api/devices/discover', {
        method: 'POST', headers: {'Content-Type': 'application/json'},
        body: JSON.stringify({})});
      const data = await response.json();
      if (!response.ok) throw new Error(data.error || `HTTP ${response.status}`);
      message.textContent = `${data.count} device(s) answered.`;
      await refresh();
    } catch (err) {
      // Named, not swallowed. A discovery that could not run is a different
      // fact from one that found nothing.
      message.textContent = `Discovery did not run: ${err.message}`;
    } finally {
      discoverBtn.disabled = false;
    }
  }

  async function refresh() {
    if (!status) build();
    try {
      const response = await fetch('/api/devices/tree');
      if (!response.ok) throw new Error(`HTTP ${response.status}`);
      snapshot = await response.json();
      render();
    } catch (err) {
      if (status) {
        status.textContent = `Could not read the device tree: ${err.message}`;
        status.dataset.state = 'bad';
      }
    }
  }

  window.addEventListener('agentmux:ready', () => {
    api = window.AGENTMUX;
    // Slower than the tag table: this merges what other panels have already
    // gathered and nothing here is time-critical.
    api.registerCard('iiot', refresh, 4000);
  }, {once: true});
})();

File: dashboard/devicetree.py
"""One device tree from several kinds of knowing, without pretending they are equal.

WHAT THIS MERGES, and why the origins have to survive the merge:

  segment-scan      netscan.py. A TCP connect answered on port 44818, and the MAC
                    resolved to a vendor through an OUI table. That is an
                    INFERENCE: something is listening, and the NIC was sold by
                    Rockwell. It is not the device telling you what it is.
  cip-listidentity  rockwell.discover(). A UDP broadcast the device ANSWERED,
                    with its own vendor, product code, revision, serial and
                    state. The device said so.
  opcua-endpoints   getEndpoints(). Also self-reported, and also optional - the
                    embedded server is firmware- and SKU-dependent (ADR-0019).
  promoted          a row the operator already saved in cc.db.

Collapsing those into one "vendor" field would be the same class of lie as an
un-aged value on the tag table: a screen that looks equally confident about a
guess and a statement. So every row carries `sources`, every field that came
from somewhere carries where, and a row that is only a port sweep says so.

DISAGREEMENT IS DATA, NOT AN ERROR TO RESOLVE. If the OUI table says one vendor
and the device says another, both are recorded and the row is flagged. Picking a
winner - newest, most specific, highest priority - is how the wrong one ends up
on screen with nothing to say it was ever in doubt. Same rule as the merged tag
table in the plan: two sources, two records, the disagreement shown.

NO NETWORK HERE. Every function is pure, so the merge can be tested against
recorded output from a real scan and a real discovery without either.
"""

import ipaddress

# Ordered by how much the device itself told us, which is the order a reader
# should trust them in. Used for display, never to pick a winner.
SOURCES = ("promoted", "cip-listidentity", "opcua-endpoints", "segment-scan")

# Ports that mean "this is worth asking a protocol question", so the tree can
# suggest the next step rather than leaving a bare open port on screen.
PROTOCOL_PORTS = {
    44818: "ethernet-ip",
    502: "modbus-tcp",
    4840: "opc-ua",
    102: "s7comm",
    1883: "mqtt",
    8883: "mqtt-tls",
    48898: "ads",
    11740: "codesys",
}


def _blank(address):
    return {
        "address": address,
        "sources": [],
        # Every one of these is None until something reports it, and each is
        # paired with the source that did. A field with no source beside it is
        # a field nobody can be asked about.
        "hostname": None,
        "mac": None,
        "vendor": None,
        "vendor_source": None,
        "ports": [],
        "identity": None,
        "endpoints": [],
        "promoted": None,
        "conflicts": [],
        "suggested": [],
    }


def _note_conflict(row, field, a_value, a_source, b_value, b_source):
    """Record that two sources disagree. Never pick one."""
    row["conflicts"].append({
        "field": field,
        "values": [{"value": a_value, "source": a_source},
                   {"value": b_value, "source": b_source}],
    })


def _add_source(row, source):
    if source not in row["sources"]:
        row["sources"].append(source)


def _sort_key(row):
    """Numeric where the address is an IP, so .10 sorts after .9."""
    try:
        return (0, int(ipaddress.ip_address(row["address"])))
    except ValueError:
        return (1, 0)


def merge(scanned=(), identified=(), endpoints=(), promoted=()):
    """Build one row per address from every source that reported it.

    Each argument is a sequence of plain dicts as its own module produces them,
    so this can be driven from recorded output with no network:

      scanned     netscan snapshot()["hosts"]
      identified  rockwell.discover()
      endpoints   [{"address":..., "endpoints":[...]}]
      promoted    cc.db devices rows
    """
    rows = {}

    def row_for(address):
        address = str(address or "").strip()
        if not address:
            return None
        if address not in rows:
            rows[address] = _blank(address)
        return rows[address]

    # ── the port sweep: something is listening, and the OUI suggests a vendor ──
    for host in scanned or ():
        row = row_for(host.get("address"))
        if row is None:
            continue
        _add_source(row, "segment-scan")
        row["hostname"] = host.get("hostname") or row["hostname"]
        row["mac"] = host.get("mac") or row["mac"]
        if host.get("vendor"):
            row["vendor"] = host["vendor"]
            row["vendor_source"] = "segment-scan (OUI)"
        row["ports"] = list(host.get("ports") or [])
        for port in row["ports"]:
            protocol = PROTOCOL_PORTS.get(port.get("port"))
            if protocol and protocol not in row["suggested"]:
                row["suggested"].append(protocol)

    # ── CIP ListIdentity: the device's own account of itself ─────────────────
    for device in identified or ():
        row = row_for(device.get("ip_address"))
        if row is None:
            continue
        _add_source(row, "cip-listidentity")
        row["identity"] = {k: device.get(k) for k in
                           ("product_name", "product_code", "vendor", "device_type",
                            "revision", "serial", "state", "status")}
        said = device.get("vendor")
        if said:
            # The device told us. If the OUI guessed something else, BOTH are
            # kept and the row is flagged - the guess is not corrected away,
            # because which one is wrong is a question for a person.
            if row["vendor"] and row["vendor"] != said:
                _note_conflict(row, "vendor", row["vendor"], row["vendor_source"],
                               said, "cip-listidentity")
            row["vendor"] = said
            row["vendor_source"] = "cip-listidentity (the device said so)"
        if "ethernet-ip" not in row["suggested"]:
            row["suggested"].append("ethernet-ip")

    # ── OPC UA endpoints: optional, and absent is normal ─────────────────────
    for entry in endpoints or ():
        row = row_for(entry.get("address"))
        if row is None:
            continue
        _add_source(row, "opcua-endpoints")
        row["endpoints"] = list(entry.get("endpoints") or [])
        if "opc-ua" not in row["suggested"]:
            row["suggested"].append("opc-ua")

    # ── what the operator already saved ──────────────────────────────────────
    for device in promoted or ():
        row = row_for(device.get("address"))
        if row is None:
            continue
        _add_source(row, "promoted")
        row["promoted"] = {k: device.get(k) for k in
                           ("id", "name", "kind", "port", "protocol")}

    return sorted(rows.values(), key=_sort_key)


def tree(rows, scan_request=None):
    """Group merged rows by subnet, for rendering.

    Grouped by the /24 the address falls in rather than by vendor or protocol,
    because that is how the person standing in front of the panel thinks about
    it: this cabinet, that cell.
    """
    groups = {}
    for row in rows:
        try:
            network = str(ipaddress.ip_network(row["address"] + "/24", strict=False))
        except ValueError:
            network = "unknown"
        groups.setdefault(network, []).append(row)

    return {
        "groups": [
            {
                "network": network,
                "devices": members,
                "count": len(members),
                # How many of these actually told us what they are, as opposed
                # to merely having a port open. The difference is the point of
                # the whole panel.
                "self_reported": sum(1 for m in members
                                     if "cip-listidentity" in m["sources"]
                                     or "opcua-endpoints" in m["sources"]),
            }
            for network, members in sorted(groups.items())
        ],
        "total": len(rows),
        "conflicts": sum(len(r["conflicts"]) for r in rows),
        # Echoed back so the panel can say what produced this, rather than
        # rendering a list with no provenance.
        "scan": scan_request,
        "sources": list(SOURCES),
    }


def promotion_for(row, name=None):
    """The cc.db `devices` row this device would be saved as.

    Deliberately uses the EXISTING schema (ccstore.py:67) - no new table and no
    new column. A device tree that needed its own store would be a second
    inventory to keep in step with the first.

    Returns None when there is nothing worth saving yet: a bare address with no
    protocol is not an inventory entry, it is a ping reply.
    """
    protocol = row["suggested"][0] if row["suggested"] else None
    if not protocol:
        return None
    port = next((p["port"] for p in row["ports"]
                 if PROTOCOL_PORTS.get(p["port"]) == protocol), None)
    identity = row.get("identity") or {}
    return {
        "name": name or identity.get("product_name") or row.get("hostname") or row["address"],
        # `kind` records HOW this was found, so an inventory entry does not lose
        # the difference between a device that identified itself and one that
        # merely had a port open.
        "kind": "cip-listidentity" if "cip-listidentity" in row["sources"] else "segment-scan",
        "address": row["address"],
        "port": port,
        "protocol": protocol,
    }
